### PR TITLE
PP-430: Test for smartWallet contract

### PR DESCRIPTION
**Pull request by AndresQuijano**
## What

- Add unit tests for SmartWallet contract

## Why

- Increase product quality

## Refs

- [PP-430](https://rsklabs.atlassian.net/browse/PP-430)

### CODE REVIEW

**Review by ironFe93 — COMMENTED** [review]
Consider my comments before merging

File: test/smartwallet/smartWallet.test.ts [hunk 1]
```diff
@@ -1,943 +1,561 @@
...
+
+            const typedRequestData = new TypedRequestData(HARDHAT_CHAIN_ID, smartWallet.address, relayRequest);            
+
+            const signature = getLocalEip712Signature(typedRequestData, privateKey);
+
+            const suffixData = getSuffixData(typedRequestData);
+
+            await expect(
+                smartWallet.verify(suffixData, relayRequest.request, signature),
+                "Verification failed"
```
[the remark this answers is not in the document]
Comment (AndresQuijano) [reply test/smartwallet/smartWallet.test.ts:197]: The test is looking for a rejection. Check the description

it('Should fail when not called by the owner'

Please let me know if I'm missing something.

File: test/smartwallet/smartWallet.test.ts [hunk 2]
```diff
@@ -1,943 +1,561 @@
...
+
+        beforeEach(async function(){
+            fakeToken = await smock.fake('ERC20');
+            fakeToken.transfer.returns(true);
+        });
+
+        it('Should initialize with the correct parameters', async function () {
+            const  {smartWallet, owner, worker} = await loadFixture(prepareFixture);
+
+            expect(await smartWallet.isInitialized(), 'Contract already initialized').to.be.false;
```
[the remark this answers is not in the document]
Comment (AndresQuijano) [reply test/smartwallet/smartWallet.test.ts]: Agree. Deleted.

File: test/smartwallet/smartWallet.test.ts [hunk 3]
```diff
@@ -1,943 +1,561 @@
...
+            SignTypedDataVersion.V4
+        );
+
+        const messageSize = Object.keys(typedRequestData.message).length;
+
+        return '0x'+(encoded.slice(messageSize * ONE_FIELD_IN_BYTES)).toString('hex');
+    }
+
+    describe('Function initialize()', function(){
+        let fakeToken:FakeContract;
```
[the remark this answers is not in the document]
Comment (AndresQuijano) [reply test/smartwallet/smartWallet.test.ts:77]: Implemented. Thanks.

File: test/smartwallet/smartWallet.test.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (ironFe93) [reply test/smartwallet/smartWallet.test.ts:192]: I did no see the "not". Everything's fine then.

**Review by ironFe93 — APPROVED** [review]
Looking good

File: test/utils/EIP712Utils.ts [hunk 4]
```diff
@@ -0,0 +1,131 @@
+import {
+    MessageTypeProperty,
+    MessageTypes,
+    TypedMessage,
+    signTypedData,
+    SignTypedDataVersion
+} from '@metamask/eth-sig-util';
+
+export interface ForwardRequest {
```
Comment (franciscotobar) [inline test/utils/EIP712Utils.ts:9]: We had a conversation with Juraj during a daily about where we can get or  define it. Its suppose that we need to get it from the typechain-types folder.

File: test/smartwallet/smartWallet.test.ts [hunk 5]
```diff
@@ -1,943 +1,537 @@
...
+
+        async function prepareFixture(){
+            const smartWalletFactory =await hardhat.getContractFactory('SmartWallet');
+            const smartWallet = await smartWalletFactory.deploy();
+            const [owner, worker, utilSigner] = await hardhat.getSigners();
```
Comment (franciscotobar) [inline test/smartwallet/smartWallet.test.ts:82]: The signers does not change during the executions, it should not be inside a fixture.

**Comment by franciscotobar** [comment]
I tried to run it locally but 4 test failed.

File: test/smartwallet/smartWallet.test.ts [hunk 6]
```diff
@@ -1,943 +1,537 @@
...
+
+            await expect(
+                smartWallet.initialize(owner.address, fakeToken.address, worker.address, 10, 400000),
+                'Deployment should be reverted'
+            ).to.be.revertedWith('Unable to pay for deployment');
+        });
+    });
+
+    describe('Function verify()', function(){
+        const HARDHAT_CHAIN_ID = 31337;
```
Comment (franciscotobar) [inline test/smartwallet/smartWallet.test.ts:146]: We should get the chain ID from the blockchain itself.

File: test/smartwallet/smartWallet.test.ts [hunk 7]
```diff
@@ -1,943 +1,537 @@
...
+        async function prepareFixture(){
+            const smartWalletFactory =await hardhat.getContractFactory('SmartWallet');
+            const smartWallet = await smartWalletFactory.deploy();
+            const [, worker, utilSigner] = await hardhat.getSigners();
+
+            const provider = hardhat.getDefaultProvider();
+            externalWallet =  hardhat.Wallet.createRandom();
+            externalWallet.connect(provider);
+
+            privateKey = Buffer.from(externalWallet.privateKey.substring(2, 66), 'hex');
```
Comment (franciscotobar) [inline test/smartwallet/smartWallet.test.ts:164]: Im not sure if this should be inside the fixture or in the beforeEach itself.

File: test/smartwallet/smartWallet.test.ts [hunk 8]
```diff
@@ -1,943 +1,537 @@
...
+                from: ZERO_ADDRESS,
+                to: ZERO_ADDRESS,
+                tokenContract: ZERO_ADDRESS,
+                value: '0',
+                gas: '10000',
+                nonce: '0',
+                tokenAmount: '0',
+                tokenGas: '50000',
+                data: '0x'
+            } as ForwardRequest,
```
Comment (franciscotobar) [inline test/smartwallet/smartWallet.test.ts:42]: the as ForwardRequest is not necessary

File: test/smartwallet/smartWallet.test.ts [hunk 9]
```diff
@@ -1,943 +1,537 @@
...
+                tokenGas: '50000',
+                data: '0x'
+            } as ForwardRequest,            
+            relayData:{
+                gasPrice: '1',
+                relayWorker: ZERO_ADDRESS,
+                callForwarder: ZERO_ADDRESS,
+                callVerifier: ZERO_ADDRESS
+            } as RelayData
+        } as RelayRequest
```
Comment (franciscotobar) [inline test/smartwallet/smartWallet.test.ts:49]: the `as RelayRequest` is not necessary

File: test/smartwallet/smartWallet.test.ts [hunk 10]
```diff
@@ -1,943 +1,537 @@
...
+                tokenAmount: '0',
+                tokenGas: '50000',
+                data: '0x'
+            } as ForwardRequest,            
+            relayData:{
+                gasPrice: '1',
+                relayWorker: ZERO_ADDRESS,
+                callForwarder: ZERO_ADDRESS,
+                callVerifier: ZERO_ADDRESS
+            } as RelayData
```
Comment (franciscotobar) [inline test/smartwallet/smartWallet.test.ts:48]: the `as RelayData` is not necessary

File: test/smartwallet/smartWallet.test.ts (hunk 6 above)
Comment (AndresQuijano) [reply test/smartwallet/smartWallet.test.ts:146]: As explained [here](https://stackoverflow.com/questions/66923541/get-chain-name-in-hardhat-not-chainid) and other posts on the Internet, there is not a simple way to obtain the chain Id from the blockchain. The mechanisms required implies external libraries and configurations. I think the work required is not worth it.

File: test/utils/EIP712Utils.ts (hunk 4 above)
Comment (AndresQuijano) [reply test/utils/EIP712Utils.ts:9]: Refactored!

File: test/smartwallet/smartWallet.test.ts (hunk 5 above)
Comment (AndresQuijano) [reply test/smartwallet/smartWallet.test.ts]: Done.

File: test/smartwallet/smartWallet.test.ts (hunk 7 above)
Comment (AndresQuijano) [reply test/smartwallet/smartWallet.test.ts]: I moved this to the beforeEach.

File: test/smartwallet/smartWallet.test.ts (hunk 8 above)
Comment (AndresQuijano) [reply test/smartwallet/smartWallet.test.ts]: Deleted.

File: test/smartwallet/smartWallet.test.ts (hunk 9 above)
Comment (AndresQuijano) [reply test/smartwallet/smartWallet.test.ts]: Deleted.

File: test/smartwallet/smartWallet.test.ts (hunk 10 above)
Comment (AndresQuijano) [reply test/smartwallet/smartWallet.test.ts]: Deleted.

File: test/smartwallet/smartWallet.test.ts [hunk 11]
```diff
@@ -1,943 +1,519 @@
...
+const ONE_FIELD_IN_BYTES = 32;
+
+type ForwardRequest = IForwarder.ForwardRequestStruct;
+type RelayData = EnvelopingTypes.RelayDataStruct;
+type RelayRequest = EnvelopingTypes.RelayRequestStruct;
+
+describe('SmartWallet', function(){
+    function createRequest(
+        request: Partial<ForwardRequest>,
+        relayData: Partial<RelayData>
```
Comment (franciscotobar) [inline test/smartwallet/smartWallet.test.ts:30]: can we do the `relayData: Partial<RelayData>` optional?

**Comment by franciscotobar** [comment]
The only missing part is the recover method that we talked about.

**Review by franciscotobar — APPROVED** [review]
LGTM

File: test/smartwallet/smartWallet.test.ts (hunk 11 above)
Comment (AndresQuijano) [reply test/smartwallet/smartWallet.test.ts]: Done.

**Pull request merged**